### PR TITLE
Deprecated \tl_sort:nN for \tl_sort_use:nN

### PR DESCRIPTION
**Pull request by josephwright**
See #914

### CODE REVIEW

**Review by FrankMittelbach — APPROVED** [review]

File: l3kernel/doc/l3obsolete.txt [hunk 1]
```diff
@@ -57,6 +57,7 @@ Function                            Date deprecated
 \tl_lower_case:nn                        2020-01-03
 \tl_mixed_case:n                         2020-01-03
 \tl_mixed_case:nn                        2020-01-03
+\tl_sort:nN                              2023-05-19
```
Comment (mrpiggi) [inline l3kernel/doc/l3obsolete.txt:60]: `\tl_reverse:n` is missing

File: l3kernel/CHANGELOG.md [hunk 2]
```diff
@@ -13,10 +13,16 @@ this project uses date-based 'snapshot' version identifiers.
 - `\coffin_(g)reset_poles:N`
 - `\int_if_zero:n(TF)`
 - `\str_mdfive_hash:n`
+-  `\tl_reverse_use:n` (was `\tl_reverse:n`)
+-  `\tl_sort_use:nN` (was `\tl_sort:nN`)
```
Comment (muzimuzhi) [inline l3kernel/CHANGELOG.md:17]: extra space character between `-` and `` ` ``
```suggestion
- `\tl_reverse_use:n` (was `\tl_reverse:n`)
- `\tl_sort_use:nN` (was `\tl_sort:nN`)
```

**Review by muzimuzhi — COMMENTED** [review]
minor styling things

File: l3kernel/testfiles/m3sort002.lvt
[the remark this answers is not in the document]
Comment (josephwright) [reply l3kernel/testfiles/m3sort002.lvt:1]: Done

**Comment by josephwright** [comment]
Comments from anyone other than @FrankMittelbach?

**Comment by blefloch** [comment]
We should consider the cases of `\tl_reverse_items:n` and `\tl_rescan:nn`.

**Comment by josephwright** [comment]
I've looked at `\tl_reverse_items:n` and `\tl_rescan:nn` and 'm not sure they need a change. For both `\tl_reverse:n` and `\tl_sort:nn` the issue was that there is an `:N` version that's not expandable and not equal to the `V`-type variant. However, we have neither `\tl_rescan:Nn` nor `\tl_reverse_items:N`. So I'm not seeing a need to alter them.

**Comment by josephwright** [comment]
Are we happy with this change? I've rebased, etc., so we are clean to merge.

**Review by muzimuzhi — CHANGES_REQUESTED** [review]
Another rebasing is required.

PS: `expl3-intro.tex` mentioned `\tl_reverse:n` twice,
```
./articles/expl3-intro.tex:219:With a macro like |\tl_reverse:n|, we know simply from the macro name that
./articles/expl3-intro.tex:220:it takes a single argument; |\tl_reverse:n{abc}|, which
```
Although `expl3-intro.tex` was not maintained for a long time...

**Comment by josephwright** [comment]
> Another rebasing is required.
> 
> PS: `expl3-intro.tex` mentioned `\tl_reverse:n` twice,
> 
> ```
> ./articles/expl3-intro.tex:219:With a macro like |\tl_reverse:n|, we know simply from the macro name that
> ./articles/expl3-intro.tex:220:it takes a single argument; |\tl_reverse:n{abc}|, which
> ```
> 
> Although `expl3-intro.tex` was not maintained for a long time...

This is the source for a published TUGBoat article: as such, we will never update it again.

File: l3kernel/CHANGELOG.md [hunk 3]
```diff
@@ -10,13 +10,17 @@ this project uses date-based 'snapshot' version identifiers.
 ### Added
 - `\text_titlecase_all:n(n)`
 - `\token_to_catcode:N`
+-  `\tl_reverse_use:n` (was `\tl_reverse:n`)
+-  `\tl_sort_use:nN` (was `\tl_sort:nN`)
```
[the remark this answers is not in the document]
Comment (josephwright) [reply l3kernel/CHANGELOG.md:14]: Corrected

File: l3kernel/l3sort.dtx [hunk 4]
```diff
@@ -691,12 +691,12 @@
 % greater or equal to |#4|, 3.~comparison, 4.~pivot, 5.~next item to
 % test.  If |#5| is the tail of the list, call \cs{tl_sort:nN} on |#1|
```
[the remark this answers is not in the document]
Comment (josephwright) [reply l3kernel/l3sort.dtx:692]: Corrected

**Comment by josephwright** [comment]
@muzimuzhi Rebased again with those corrections

**Review by muzimuzhi — CHANGES_REQUESTED** [review]
Dates deprecated recorded in `l3obsolete.txt` should be updated to 2023-10-18, to be consistent with that in `l3deprecation.dtx` and `l3(tl|sort).dtx`.

File: l3kernel/doc/l3obsolete.txt [hunk 5]
```diff
@@ -70,6 +70,8 @@ Function                            Date deprecated
 \tl_lower_case:nn                        2020-01-03
 \tl_mixed_case:n                         2020-01-03
 \tl_mixed_case:nn                        2020-01-03
+\tl_reverse:n                            2023-05-22
+\tl_sort:nN                              2023-05-19
```
[the remark this answers is not in the document]
Comment (josephwright) [reply l3kernel/doc/l3obsolete.txt:74]: I'll update those if-and-when we actually merge

**Comment by blefloch** [comment]
> I've looked at `\tl_reverse_items:n` and `\tl_rescan:nn` and 'm not sure they need a change. For both `\tl_reverse:n` and `\tl_sort:nn` the issue was that there is an `:N` version that's not expandable and not equal to the `V`-type variant. However, we have neither `\tl_rescan:Nn` nor `\tl_reverse_items:N`. So I'm not seeing a need to alter them.

Thinking about it some more (sorry for the delay), I realize there could be three versions of a given token list operation:
- take some `n`-type argument, apply the operation, and use the result;
- take some `n`-type argument, apply the operation, and store the result in a variable;
- take some variable, apply the operation, and store the result back into the variable.

| n to n | n to N | N to N |
|-------|--------|-----|
| `\tl_reverse:n` |  | `\tl_reverse:N` |
| `\tl_rescan:nn` | `\tl_set_rescan:Nnn` | |
| `\tl_sort:nn` | | `\tl_sort:Nn` |
| `\tl_trim_spaces:n` | | `\tl_trim_spaces:N` |

For all other pairs of functions where both `:n` and `:N` exists, the `:N` function is equivalent to expanding once before calling the `:n` function.  For instance `\tl_to_str:N`, `\tl_count:N`, `\tl_show:N`, `\tl_map_inline:Nn`, `\tl_head:N`… I suspect that it would make more sense to add the word `set` to `\tl_reverse:N`, `\tl_sort:Nn` and `\tl_trim_spaces:N`, or add the word `use` to `\tl_reverse:n`, `\tl_sort:nn`, `\tl_trim_spaces:n` (and maybe `\tl_rescan:nn` as well).

**Review by FrankMittelbach — COMMENTED** [review]
Seeing Bruno's comments I think that I should take my earlier approval back. If one looks at the larger picture then we don't use `_use` anywhere except when we "use some register", e.g., stuff like `\dim_use:N` (N in the signature). The only other  "use" is the use module where nothing is done to the arguments other then selecting some of the argument (and possibly prehandling via "c"  etc.

All the ":n" functions that do something to the argument and then pass it back to the input stream only have the action in their name and the "use" is implicit. Of course one could add  `_use` to all of them but the convention without it very common and I think easy to understand.

Given that and given that when something is stored we normally have `_set` in the name across all modules, I think that his suggestion to add `_set` to  `\tl_sort:Nn`, either as `\tl_set_sort: Nn` (or maybe as `\tl_set_sorted:Nn` but I guess I would stay with "sort") and also to the others that save their result back into a var.

**Comment by josephwright** [comment]
> Given that and given that when something is stored we normally have `_set` in the name across all modules, I think that his suggestion to add `_set` to `\tl_sort:Nn`, either as `\tl_set_sort: Nn` (or maybe as `\tl_set_sorted:Nn` but I guess I would stay with "sort") and also to the others that save their result back into a var.

I remember suggesting `\tl_set_sorted:Nn` ages ago :)

More seriously, I think I'd prefer `\tl_set_sort:Nn` over `\tl_sort_use:n` - I'm going to close the PR as a result and will look at an alternative one.

**Comment by FrankMittelbach** [comment]
> I remember suggesting `\tl_set_sorted:Nn` ages ago :)

quite likely
> 
> More seriously, I think I'd prefer `\tl_set_sort:Nn` over `\tl_sort_use:n` - I'm going to close the PR as a result and will look at an alternative one.

Bruno's main point was the need to do that to a few other N commands as well and I don't think that was on the table before (and I think he is right there).

**Pull request closed without merging**